alacritty_terminal/src/term/mod.rs

```diff
@@ -995,6 +996,7 @@ impl<T> Term<T> {
         cursor_cell.c = c;
         cursor_cell.fg = fg;
         cursor_cell.bg = bg;
+        cursor_cell.set_underline_color(underline_color);
```

Comment:
I haven't benchmarked things yet, but this might have a big impact on performance.

The `write_at_cursor` function is **extremely** hot, so even a simple branching like in the `set_underline_color` might be problematic.

I'll just leave this comment here for now, but this might be a problem. Feel free to ignore it until everything else is done and I have run my own benchmarks though.

Reply:
I'd appreciate if you could look into optimizing this part of the code. It's most likely where the performance regression is coming from and this is a rather unfortunate implementation.

Ideally this function shouldn't have to do anything different if the underline isn't colored, but that might be tough to realize.

I could look into this myself, but since I personally don't care about this feature that's going to be at the bottom of my priority list for now.

Reply:
If performance is more important than memory maybe it is better to put the `underline_color` field in `Cell` instead of `CellExtra`.

I did a quick test in https://github.com/ayosec/alacritty/commit/bbf64848520c7b3b53614869f88f3195ca955c51 and then launched `vtebench`. In my machine, the performance after the change is better is most cases:

![output](https://user-images.githubusercontent.com/37484/105895260-35e56700-600d-11eb-9b25-d8c7adc47fcc.png)

Reply:
I got rid of 1 branch in the common case: there used to be a check for the existence of `extra` when setting underline_color to `None`, but as `extra` is dropped in `write_at_cursor`, we know this will be `None`, so we can get rid of the branch.

Reply:
Well is that enough to match master's performance?

Reply:
According to my benchmarks, which are not very consistent as you mentioned, it is only very slightly slower than master. As far as I can tell, this will only add one branch (I tried optimizing it further to only have one branch in the rust code, but the generated asm did not change), which is probably as fast as it can get.

Reply:
> which is probably as fast as it can get

"probably as fast as it can get" is not very reassuring. It doesn't just matter how often it branches, it also matters how it branches and if the assembly it generates makes sense or not.

That information is necessary to know if it's possible to store this in the dynamic data, or if maybe it has to be stored in every cell and bloat Alacritty's memory size.

There are also various tips on improving benchmarking consistency in here: https://llvm.org/docs/Benchmarking.html

Generating some alternative benchmarks for faster validation and running everything through `perf` to verify its performance probably makes sense.